game/game/movealgo.cpp

```diff
@@ -709,27 +709,31 @@ void MoveAlgo::startDive() {
   }
 
 void MoveAlgo::invalidatePhysics(const Tempest::Vec3& dp) {
-  auto         pos           = npc.position();
-  bool         validW        = false;
-  const  float fallThreshold = stepHeight();
-  const  float worldHeight   = 20000;
-  static float water;
-  const  float waterHeight   = waterRay(pos-dp, &validW);
-
-  if(validW)
-    water = waterHeight;
-  if(pos.y+fallThreshold>water) {
+  auto  pos           = npc.position();
+  float fallThreshold = stepHeight();
+  float worldHeight   = DynamicWorld::worldHeight;
+  auto  pathRay       = npc.world().physic()->ray(pos-dp,pos);
+  auto  skyRay        = npc.world().physic()->ray(Tempest::Vec3(pos.x,worldHeight,pos.z),pos);
+  bool  validW        = false;
+  waterRay(pos,&validW);
+
+  if(pathRay.hasCol) {
+    if (skyRay.hasCol)
+      npc.setPosition(skyRay.v+Tempest::Vec3(0,5*fallThreshold,0));
     setAsSwim(false);
     setAsDive(false);
     setInAir (true);
     return;
```

Comment:
`if(pathRay.hasCol)` is there to copy vanilla behavior and move the character upwards. `5*fallThreshold` is an arbitrary constant added to prevent the character getting stuck.

Reply:
This is for `K`. `F8` resets the character to the previous position if moving into a wall, but this doesn't feel natural.  `K` probably happens if non-water state is detected in vanilla. In this case swim conditions are invalidated.